modules/createBrowserHistory.js

```diff
@@ -60,19 +60,15 @@ const createBrowserHistory = (props = {}) => {
     let path = pathname + search + hash
 
     warning(
-      !(basename && path.indexOf(basename) !== 0),
+      !(basename && hasBasename(path, basename)),
       'You are attempting to use a basename on a page whose URL path does not begin ' +
       'with the basename. Expected path "' + path + '" to begin with "' + basename + '".'
     )
 
     if (basename)
-      path = stripPrefix(path, basename)
+      path = stripBasename(path, basename)
 
-    return {
-      ...parsePath(path),
-      state,
-      key
-    }
+    return createLocation(path, state, key)
```

Comment:
It's not a huge deal but I'd prefer it if hash history objects didn't have a `key` property. I think it would mislead people into thinking that we support key with hash history when we don't really.

It feels like this might be one of those changes that requires a more thorough refactoring. If that's the case, I'm happy to take it on. I've set aside some time today and tomorrow for OSS work.